{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "YAxgZo4Y4Zx3"
      },
      "source": [
        "#Trabajo Final Grupo Nº 2\n",
        "\n",
        "Rafael Perez\trafaelperezctes@gmail.com\n",
        "Olinca Ayala\tolincayalan88@gmail.com\n",
        "###@Created on Dec Sat 2 16:57:20 2023\n",
        "###@author: Grupo Nº 2"
      ],
      "id": "YAxgZo4Y4Zx3"
    },
    {
      "cell_type": "markdown",
      "id": "dbaae3b4",
      "metadata": {
        "id": "dbaae3b4"
      },
      "source": [
        "\n",
        "# Modulo guias\n",
        "\n",
        "Este cuaderno realiza la descarga y análisis de secuencias nucleotídicas de integronas de Shewanella utilizando Biopython.\n"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# @title Instalación de librerias\n",
        "import subprocess\n",
        "import ipywidgets as widgets\n",
        "from IPython.display import display\n",
        "import pkg_resources\n",
        "\n",
        "def instalar_librerias():\n",
        "    librerias = [\n",
        "        \"import_ipynb\",\n",
        "        \"nbconvert\",\n",
        "        \"ipywidgets\",\n",
        "        \"biopython\",\n",
        "        \"pandas\",\n",
        "        \"numpy\",\n",
        "        \"matplotlib\",\n",
        "        \"tqdm\",\n",
        "        \"odfpy\",\n",
        "        \"requests\",\n",
        "        \"gdown\",\n",
        "        \"mafft\",\n",
        "        \"clustalo\",\n",
        "        \"muscle\",\n",
        "        \"fasttree\",\n",
        "        \"weblogo\",\n",
        "        \"logomaker\",\n",
        "\n",
        "    ]\n",
        "\n",
        "    barra_progreso = widgets.IntProgress(\n",
        "        value=0,\n",
        "        min=0,\n",
        "        max=len(librerias),\n",
        "        bar_style='info',\n",
        "        orientation='horizontal'\n",
        "    )\n",
        "\n",
        "    etiqueta = widgets.Label(value='Iniciando instalación...')\n",
        "    display(widgets.VBox([etiqueta, barra_progreso]))\n",
        "\n",
        "    # Instalación de fasttree y mafft con apt-get (solo ejecutar en Colab)\n",
        "    subprocess.run([\"apt-get\", \"install\", \"fasttree\", \"mafft\", \"ghostscript\"])\n",
        "\n",
        "    for libreria in librerias:\n",
        "        try:\n",
        "            pkg_resources.require(libreria)\n",
        "            etiqueta.value = f\"{libreria} ya está instalado.\"\n",
        "        except pkg_resources.DistributionNotFound:\n",
        "            etiqueta.value = f\"Instalando: {libreria}...\"\n",
        "            subprocess.run([\"pip\", \"install\", libreria])\n",
        "        barra_progreso.value += 1\n",
        "\n",
        "    etiqueta.value = 'Instalación completada.'\n",
        "# Crear botón\n",
        "boton_instalar = widgets.Button(description='Instalar Librerías')\n",
        "\n",
        "# Crear área de salida\n",
        "output = widgets.Output()\n",
        "display(output)\n",
        "\n",
        "# Enlazar el botón con la función de instalación\n",
        "boton_instalar.on_click(lambda b: instalar_librerias())\n",
        "\n",
        "# Mostrar botón\n",
        "display(boton_instalar)\n",
        "\n",
        "#instalar_librerias()\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 49,
          "referenced_widgets": [
            "fab05358eeda49878e4c60ebb7c5d3e4",
            "47a736ff96d04839a63d4707b8dbc9f0",
            "84fcecaa4f194245aaef610ec099655c",
            "e2bba9d4197c4c84ba35e5944acbdaca",
            "de922e72fe0d4bec8cb991b82fd66729"
          ]
        },
        "id": "rlMGb_QOAR9K",
        "outputId": "6332d974-f3dc-467e-86f7-5cb410d21af7"
      },
      "execution_count": 1,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "Output()"
            ],
            "application/vnd.jupyter.widget-view+json": {
              "version_major": 2,
              "version_minor": 0,
              "model_id": "fab05358eeda49878e4c60ebb7c5d3e4"
            }
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "Button(description='Instalar Librerías', style=ButtonStyle())"
            ],
            "application/vnd.jupyter.widget-view+json": {
              "version_major": 2,
              "version_minor": 0,
              "model_id": "84fcecaa4f194245aaef610ec099655c"
            }
          },
          "metadata": {}
        }
      ],
      "id": "rlMGb_QOAR9K"
    },
    {
      "cell_type": "markdown",
      "source": [
        "1. Selección de la Región Objetivo:\n",
        "\n",
        "- Identifica la región del genoma que deseas editar. Esto podría ser un gen específico o una secuencia de interés.\n",
        "\n",
        "- Asegúrate de que esta región sea accesible para el sistema CRISPR/Cas9 y no esté oculta o inaccesible en la estructura cromosómica.\n",
        "\n",
        "2. Diseño de la Secuencia Guía (sgRNA):\n",
        "\n",
        "- La secuencia guía (sgRNA) debe ser complementaria a la región objetivo en el genoma para guiar la enzima Cas9 al sitio correcto.\n",
        "- La longitud típica de una sgRNA es de aproximadamente 20 nucleótidos.\n",
        "Para el sistema CRISPR/Cas9 común (SpCas9), la sgRNA debe estar adyacente a una secuencia PAM (Protospacer Adjacent Motif). Para SpCas9, el PAM es típicamente 'NGG'.\n",
        "- Evita regiones con alta homología a otras partes del genoma para minimizar los efectos fuera del blanco (off-target effects)."
      ],
      "metadata": {
        "id": "ho1G7kjNsvmo"
      },
      "id": "ho1G7kjNsvmo"
    },
    {
      "cell_type": "code",
      "source": [
        "# @title main\n",
        "\n",
        "def main():\n",
        "  instalar_librerias()\n",
        "\n",
        "if __name__ == \"__main__\":\n",
        "    main()\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 81,
          "referenced_widgets": [
            "3f0c138212804cd2b041519ac1ceea39",
            "8c8215497630446e8149ad684a1aed7c",
            "c554ae938f174612b2f552bae358fd1f",
            "c7aa1fcc766e4adcb7225b6e9e307ff9",
            "19c877fcda3c4c6b80f69f01df225fb8",
            "a3660ac834fd41238cdae2566b7ebfc1",
            "a7ee45cb347b4689b233220643695c73",
            "41505775e0284e18810c991d290f047b"
          ]
        },
        "id": "00i9DQuFpnWA",
        "outputId": "547fff49-bbfe-4789-e691-11c5d15266ac"
      },
      "id": "00i9DQuFpnWA",
      "execution_count": 2,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "VBox(children=(Label(value='Iniciando instalación...'), IntProgress(value=0, bar_style='info', max=17)))"
            ],
            "application/vnd.jupyter.widget-view+json": {
              "version_major": 2,
              "version_minor": 0,
              "model_id": "3f0c138212804cd2b041519ac1ceea39"
            }
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# @title crea guias\n",
        "# crea 10 guias por secuencia dada una matriz pam\n",
        "\n",
        "import Bio\n",
        "from Bio import SeqIO\n",
        "from Bio.SeqUtils import gc_fraction, MeltingTemp as mt\n",
        "import pandas as pd\n",
        "\n",
        "# Variable global para la matriz PAM\n",
        "PAM_SEQUENCE = 'GG'\n",
        "\n",
        "# Función para leer secuencias desde archivos FASTA\n",
        "def read_fasta_files(file_list):\n",
        "    sequences = {}\n",
        "    for file in file_list:\n",
        "        with open(file, \"r\") as fasta_file:\n",
        "            for record in SeqIO.parse(fasta_file, \"fasta\"):\n",
        "                sequences[file] = str(record.seq)\n",
        "    return sequences\n",
        "\n",
        "# Función para generar guías de ARN para CRISPR-Cas9\n",
        "def generate_guides(dna_seq, pam_sequence=PAM_SEQUENCE):\n",
        "    guides = []\n",
        "    for i in range(len(dna_seq) - 22):\n",
        "        seq = dna_seq[i:i+23]\n",
        "        if seq.endswith(pam_sequence):\n",
        "            guide = seq[:20]\n",
        "            guides.append(guide)\n",
        "    return guides\n",
        "\n",
        "# Función para evaluar eficiencia de corte y especificidad de las guías\n",
        "def score_guides(guides, dna_seq):\n",
        "    scored_guides = []\n",
        "    for guide in guides:\n",
        "        gc_content = gc_fraction(Bio.Seq.Seq(guide)) * 100  # Usando gc_fraction\n",
        "        tm = mt.Tm_NN(Bio.Seq.Seq(guide))\n",
        "        poly_t = guide.count('TTTT')\n",
        "        specificity = dna_seq.count(guide) == 1\n",
        "        score = (gc_content / 100 * 40) + (tm / 85 * 60) - (poly_t * 5) + (specificity * 20)\n",
        "        scored_guides.append((guide, score, gc_content, tm, specificity))\n",
        "    return scored_guides\n",
        "\n",
        "# Función para seleccionar las mejores guías según su puntuación\n",
        "def select_best_guides(scored_guides):\n",
        "    return sorted(scored_guides, key=lambda x: x[1], reverse=True)[:10]\n",
        "\n",
        "# Función para seleccionar las mejores guías según su puntuación\n",
        "def process_fasta_files(file_list):\n",
        "    sequences = read_fasta_files(file_list)\n",
        "    all_guides_df = pd.DataFrame()\n",
        "\n",
        "    for file, seq in sequences.items():\n",
        "        guides = generate_guides(seq)\n",
        "        scored_guides = score_guides(guides, seq)\n",
        "        best_guides = select_best_guides(scored_guides)\n",
        "\n",
        "        # Crear DataFrame para las guías\n",
        "        guides_df = pd.DataFrame(best_guides, columns=['Guide', 'Score', 'GC%', 'Tm', 'Specificity'])\n",
        "        guides_df['File'] = file\n",
        "\n",
        "        # Agregar la fila separadora\n",
        "        separator = pd.DataFrame([[\"-\"*20, \"-\"*5, \"-\"*5, \"-\"*5, \"-\"*5, \"\"]], columns=guides_df.columns)\n",
        "\n",
        "        # Usar pandas.concat en lugar de DataFrame.append\n",
        "        all_guides_df = pd.concat([all_guides_df, guides_df, separator], ignore_index=True)\n",
        "\n",
        "    # Eliminar la última fila separadora\n",
        "    return all_guides_df.iloc[:-1]\n",
        "\n",
        "#all_guides_df = process_fasta_files(['U12441.2.fasta', 'AY509004.fasta'])\n",
        "#all_guides_df\n"
      ],
      "metadata": {
        "id": "Z6IRF--TvI4a"
      },
      "id": "Z6IRF--TvI4a",
      "execution_count": 3,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# @title imprimir comparar guias\n",
        "\n",
        "import matplotlib.pyplot as plt\n",
        "\n",
        "# Suponiendo que 'all_guides_df' es el DataFrame resultante de tu función 'process_fasta_files'\n",
        "def plot_guide_comparisons(df):\n",
        "    # Filtrar las filas que no son guías de ARN (es decir, las filas de separación)\n",
        "    df_filtered = df[df['Guide'].apply(lambda x: x.isalpha())]\n",
        "\n",
        "    # Crear subplots para cada característica\n",
        "    fig, axes = plt.subplots(nrows=1, ncols=3, figsize=(15, 5))\n",
        "\n",
        "    # Contenido de GC\n",
        "    axes[0].scatter(df_filtered['File'], df_filtered['GC%'])\n",
        "    axes[0].set_title('Contenido de GC')\n",
        "    axes[0].set_xlabel('Archivo')\n",
        "    axes[0].set_ylabel('% GC')\n",
        "    axes[0].tick_params(axis='x', rotation=45)\n",
        "\n",
        "    # Temperatura de fusión (Tm)\n",
        "    axes[1].scatter(df_filtered['File'], df_filtered['Tm'])\n",
        "    axes[1].set_title('Temperatura de Fusión (Tm)')\n",
        "    axes[1].set_xlabel('Archivo')\n",
        "    axes[1].set_ylabel('Tm')\n",
        "    axes[1].tick_params(axis='x', rotation=45)\n",
        "\n",
        "    # Puntuación\n",
        "    axes[2].scatter(df_filtered['File'], df_filtered['Score'])\n",
        "    axes[2].set_title('Puntuación de la Guía')\n",
        "    axes[2].set_xlabel('Archivo')\n",
        "    axes[2].set_ylabel('Puntuación')\n",
        "    axes[2].tick_params(axis='x', rotation=45)\n",
        "\n",
        "    plt.tight_layout()\n",
        "    plt.show()\n",
        "\n",
        "#all_guides_df = process_fasta_files(['U12441.2.fasta', 'AY509004.fasta'])  # Ejemplo de llamada a la función\n",
        "#plot_guide_comparisons(all_guides_df)\n"
      ],
      "metadata": {
        "id": "YQ-9Y2qpBAH8"
      },
      "id": "YQ-9Y2qpBAH8",
      "execution_count": 4,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# @title verificar guias 0\n",
        "\"\"\"\n",
        "import Bio\n",
        "from Bio import SeqIO\n",
        "from Bio.SeqUtils import GC\n",
        "import pandas as pd\n",
        "import matplotlib.pyplot as plt\n",
        "import numpy as np\n",
        "import random\n",
        "\n",
        "# Función para seleccionar la mejor guía para cada archivo\n",
        "def select_best_guides_per_file0(df):\n",
        "    df_filtered = df[df['Guide'].apply(lambda x: x.isalpha())]\n",
        "    return df_filtered.sort_values(by=['File', 'Score'], ascending=[True, False]).groupby('File').head(1)\n",
        "\n",
        "# Función para transformar bacterias simulando ediciones CRISPR/Cas9\n",
        "def transform_bacteria0(guide, sequence):\n",
        "    mutations = []\n",
        "    start = sequence.find(guide)\n",
        "    if start != -1:\n",
        "        end = start + len(guide)\n",
        "        mutation_type = random.choice([\"deletion\", \"insertion\", \"substitution\"])\n",
        "        if mutation_type == \"deletion\":\n",
        "            mutated_seq = sequence[:start] + sequence[end:]\n",
        "            mutations.append((mutation_type, start, sequence[start:end], \"\"))\n",
        "        elif mutation_type == \"insertion\":\n",
        "            insertion_seq = \"ATCG\"\n",
        "            mutated_seq = sequence[:start] + insertion_seq + sequence[start:]\n",
        "            mutations.append((mutation_type, start, \"\", insertion_seq))\n",
        "        elif mutation_type == \"substitution\":\n",
        "            substitution_seq = \"ATCG\"\n",
        "            mutated_seq = sequence[:start] + substitution_seq + sequence[end:]\n",
        "            mutations.append((mutation_type, start, sequence[start:end], substitution_seq))\n",
        "        return mutated_seq, mutations\n",
        "    else:\n",
        "        return sequence, []\n",
        "\n",
        "# Función para analizar y visualizar los resultados\n",
        "def analyze_and_visualize0(file_name, original_sequence, edited_sequence, mutations):\n",
        "    print(f\"\\nAnálisis de {file_name}:\")\n",
        "    # Análisis de proteínas y codones\n",
        "    old_protein = original_sequence.translate()\n",
        "    new_protein = edited_sequence.translate()\n",
        "    old_codon_counts = count_codons(original_sequence)\n",
        "    new_codon_counts = count_codons(edited_sequence)\n",
        "\n",
        "    print(\"Proteína antigua: \", old_protein[:50], \"...\")\n",
        "    print(\"Proteína nueva: \", new_protein[:50], \"...\")\n",
        "    #print(\"Codones antes de la edición: \", old_codon_counts)\n",
        "    #print(\"Codones después de la edición: \", new_codon_counts)\n",
        "\n",
        "    # Visualización gráfica de los cambios\n",
        "    visualize_changes(old_codon_counts, new_codon_counts, original_sequence, edited_sequence)\n",
        "\n",
        "# Función para contar codones\n",
        "def count_codons0(seq):\n",
        "    return pd.Series([str(seq[i:i+3]) for i in range(0, len(seq), 3)]).value_counts()\n",
        "\n",
        "# Función para visualizar los cambios\n",
        "def visualize_changes0(old_codon_counts, new_codon_counts, old_seq, new_seq):\n",
        "    plt.figure(figsize=(10, 5))\n",
        "    plt.bar(old_codon_counts.index, old_codon_counts.values, alpha=0.5, label='Antes')\n",
        "    plt.bar(new_codon_counts.index, new_codon_counts.values, alpha=0.5, label='Después')\n",
        "    plt.xticks(rotation=90)\n",
        "    plt.ylabel('Frecuencia')\n",
        "    plt.title('Comparación de Codones Antes y Después de la Edición Genética')\n",
        "    plt.legend()\n",
        "    plt.tight_layout()\n",
        "    plt.show()\n",
        "\n",
        "    # Visualización del contenido de GC\n",
        "    gc_content_before = GC(old_seq)\n",
        "    gc_content_after = GC(new_seq)\n",
        "    plt.figure()\n",
        "    plt.bar([\"Antes\", \"Después\"], [gc_content_before, gc_content_after], color=['blue', 'orange'])\n",
        "    plt.ylabel('Contenido de GC (%)')\n",
        "    plt.title('Contenido de GC Antes y Después de la Edición Genética')\n",
        "    plt.show()\n",
        "\n",
        "# Proceso principal para cada archivo\n",
        "def process_files0(df, file_list):\n",
        "    for file_name in file_list:\n",
        "        guide = df[df['File'] == file_name]['Guide'].iloc[0]\n",
        "        with open(file_name, \"r\") as fasta_file:\n",
        "            original_sequence = SeqIO.read(fasta_file, \"fasta\").seq\n",
        "            edited_sequence, mutations = transform_bacteria(guide, original_sequence)\n",
        "\n",
        "            # Análisis y visualización\n",
        "            analyze_and_visualize(file_name, original_sequence, edited_sequence, mutations)\n",
        "\n",
        "# Ejemplo de uso\n",
        "fasta_files = ['U12441.2.fasta', 'AY509004.fasta']  # Lista de tus archivos FASTA\n",
        "all_guides_df = process_fasta_files(fasta_files)\n",
        "best_guides_df = select_best_guides_per_file0(all_guides_df)\n",
        "process_files0(best_guides_df, fasta_files)\n",
        "\"\"\""
      ],
      "metadata": {
        "id": "E7fROoMtb8uU",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 191
        },
        "outputId": "65acd220-eef3-4869-e470-817f18c1d4d8"
      },
      "id": "E7fROoMtb8uU",
      "execution_count": 29,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "'\\nimport Bio\\nfrom Bio import SeqIO\\nfrom Bio.SeqUtils import GC\\nimport pandas as pd\\nimport matplotlib.pyplot as plt\\nimport numpy as np\\nimport random\\n\\n# Función para seleccionar la mejor guía para cada archivo\\ndef select_best_guides_per_file0(df):\\n    df_filtered = df[df[\\'Guide\\'].apply(lambda x: x.isalpha())]\\n    return df_filtered.sort_values(by=[\\'File\\', \\'Score\\'], ascending=[True, False]).groupby(\\'File\\').head(1)\\n\\n# Función para transformar bacterias simulando ediciones CRISPR/Cas9\\ndef transform_bacteria0(guide, sequence):\\n    mutations = []\\n    start = sequence.find(guide)\\n    if start != -1:\\n        end = start + len(guide)\\n        mutation_type = random.choice([\"deletion\", \"insertion\", \"substitution\"])\\n        if mutation_type == \"deletion\":\\n            mutated_seq = sequence[:start] + sequence[end:]\\n            mutations.append((mutation_type, start, sequence[start:end], \"\"))\\n        elif mutation_type == \"insertion\":\\n            insertion_seq = \"ATCG\"\\n            mutated_seq = sequence[:start] + insertion_seq + sequence[start:]\\n            mutations.append((mutation_type, start, \"\", insertion_seq))\\n        elif mutation_type == \"substitution\":\\n            substitution_seq = \"ATCG\"\\n            mutated_seq = sequence[:start] + substitution_seq + sequence[end:]\\n            mutations.append((mutation_type, start, sequence[start:end], substitution_seq))\\n        return mutated_seq, mutations\\n    else:\\n        return sequence, []\\n\\n# Función para analizar y visualizar los resultados\\ndef analyze_and_visualize0(file_name, original_sequence, edited_sequence, mutations):\\n    print(f\"\\nAnálisis de {file_name}:\")\\n    # Análisis de proteínas y codones\\n    old_protein = original_sequence.translate()\\n    new_protein = edited_sequence.translate()\\n    old_codon_counts = count_codons(original_sequence)\\n    new_codon_counts = count_codons(edited_sequence)\\n\\n    print(\"Proteína antigua: \", old_protein[:50], \"...\")\\n    print(\"Proteína nueva: \", new_protein[:50], \"...\")\\n    #print(\"Codones antes de la edición: \", old_codon_counts)\\n    #print(\"Codones después de la edición: \", new_codon_counts)\\n\\n    # Visualización gráfica de los cambios\\n    visualize_changes(old_codon_counts, new_codon_counts, original_sequence, edited_sequence)\\n\\n# Función para contar codones\\ndef count_codons0(seq):\\n    return pd.Series([str(seq[i:i+3]) for i in range(0, len(seq), 3)]).value_counts()\\n\\n# Función para visualizar los cambios\\ndef visualize_changes0(old_codon_counts, new_codon_counts, old_seq, new_seq):\\n    plt.figure(figsize=(10, 5))\\n    plt.bar(old_codon_counts.index, old_codon_counts.values, alpha=0.5, label=\\'Antes\\')\\n    plt.bar(new_codon_counts.index, new_codon_counts.values, alpha=0.5, label=\\'Después\\')\\n    plt.xticks(rotation=90)\\n    plt.ylabel(\\'Frecuencia\\')\\n    plt.title(\\'Comparación de Codones Antes y Después de la Edición Genética\\')\\n    plt.legend()\\n    plt.tight_layout()\\n    plt.show()\\n\\n    # Visualización del contenido de GC\\n    gc_content_before = GC(old_seq)\\n    gc_content_after = GC(new_seq)\\n    plt.figure()\\n    plt.bar([\"Antes\", \"Después\"], [gc_content_before, gc_content_after], color=[\\'blue\\', \\'orange\\'])\\n    plt.ylabel(\\'Contenido de GC (%)\\')\\n    plt.title(\\'Contenido de GC Antes y Después de la Edición Genética\\')\\n    plt.show()\\n\\n# Proceso principal para cada archivo\\ndef process_files0(df, file_list):\\n    for file_name in file_list:\\n        guide = df[df[\\'File\\'] == file_name][\\'Guide\\'].iloc[0]\\n        with open(file_name, \"r\") as fasta_file:\\n            original_sequence = SeqIO.read(fasta_file, \"fasta\").seq\\n            edited_sequence, mutations = transform_bacteria(guide, original_sequence)\\n\\n            # Análisis y visualización\\n            analyze_and_visualize(file_name, original_sequence, edited_sequence, mutations)\\n\\n# Ejemplo de uso\\nfasta_files = [\\'U12441.2.fasta\\', \\'AY509004.fasta\\']  # Lista de tus archivos FASTA\\nall_guides_df = process_fasta_files(fasta_files)\\nbest_guides_df = select_best_guides_per_file0(all_guides_df)\\nprocess_files0(best_guides_df, fasta_files)\\n'"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "string"
            }
          },
          "metadata": {},
          "execution_count": 29
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# @title verificar guias 1\n",
        "\n",
        "import Bio\n",
        "from Bio import SeqIO\n",
        "from Bio.SeqUtils import GC\n",
        "import pandas as pd\n",
        "import matplotlib.pyplot as plt\n",
        "import numpy as np\n",
        "import random\n",
        "from Bio.Seq import Seq\n",
        "\n",
        "def select_best_guides_per_file(df):\n",
        "    # Verificar si el DataFrame está vacío\n",
        "    if df.empty:\n",
        "        print(\"Error: El DataFrame proporcionado está vacío.\")\n",
        "        return pd.DataFrame()\n",
        "\n",
        "    # Filtrar el DataFrame para obtener solo guías válidas\n",
        "    df_filtered = df[df['Guide'].apply(lambda x: x.isalpha())]\n",
        "\n",
        "    # Verificar si hay guías válidas\n",
        "    if df_filtered.empty:\n",
        "        print(\"Error: No se encontraron guías válidas en el DataFrame.\")\n",
        "        return pd.DataFrame()\n",
        "\n",
        "    # Si hay guías válidas, retornar las mejores guías\n",
        "    return df_filtered.sort_values(by=['File', 'Score'], ascending=[True, False]).groupby('File').head(1)\n",
        "\n",
        "\n",
        "# Función para transformar bacterias simulando ediciones CRISPR/Cas9\n",
        "def transform_bacteria(guide, sequence):\n",
        "    mutations = []\n",
        "    start = sequence.find(guide)\n",
        "    if start != -1:\n",
        "        end = start + len(guide)\n",
        "        mutation_type = random.choice([\"deletion\", \"insertion\", \"substitution\"])\n",
        "        if mutation_type == \"deletion\":\n",
        "            mutated_seq = sequence[:start] + sequence[end:]\n",
        "            mutations.append((mutation_type, start, sequence[start:end], \"\"))\n",
        "        elif mutation_type == \"insertion\":\n",
        "            insertion_seq = \"ATCG\"\n",
        "            mutated_seq = sequence[:start] + insertion_seq + sequence[start:]\n",
        "            mutations.append((mutation_type, start, \"\", insertion_seq))\n",
        "        elif mutation_type == \"substitution\":\n",
        "            substitution_seq = \"ATCG\"\n",
        "            mutated_seq = sequence[:start] + substitution_seq + sequence[end:]\n",
        "            mutations.append((mutation_type, start, sequence[start:end], substitution_seq))\n",
        "        return mutated_seq, mutations\n",
        "    else:\n",
        "        return sequence, []\n",
        "def analyze_and_visualize(file_name, original_sequence, edited_sequence, mutations):\n",
        "    print(f\"\\nAnálisis de {file_name}:\")\n",
        "\n",
        "    # Convertir las secuencias de str a objetos Seq\n",
        "    original_seq_obj = Seq(original_sequence)\n",
        "    edited_seq_obj = Seq(edited_sequence)\n",
        "\n",
        "    # Análisis de proteínas y codones\n",
        "    old_protein = original_seq_obj.translate()\n",
        "    new_protein = edited_seq_obj.translate()\n",
        "\n",
        "    print(\"Proteína antigua: \", old_protein[:50], \"...\")\n",
        "    print(\"Proteína nueva: \", new_protein[:50], \"...\")\n",
        "\n",
        "    old_codon_counts = count_codons(original_seq_obj)\n",
        "    new_codon_counts = count_codons(edited_seq_obj)\n",
        "\n",
        "    # Visualización gráfica de los cambios\n",
        "    visualize_changes(old_codon_counts, new_codon_counts, original_sequence, edited_sequence)\n",
        "\n",
        "def count_codons(seq):\n",
        "    return pd.Series([str(seq[i:i+3]) for i in range(0, len(seq) - len(seq) % 3, 3)]).value_counts()\n",
        "\n",
        "def visualize_changes(old_codon_counts, new_codon_counts, old_seq, new_seq):\n",
        "    plt.figure(figsize=(10, 5))\n",
        "    plt.bar(old_codon_counts.index, old_codon_counts.values, alpha=0.5, label='Antes')\n",
        "    plt.bar(new_codon_counts.index, new_codon_counts.values, alpha=0.5, label='Después', color='orange')\n",
        "    plt.xticks(rotation=90)\n",
        "    plt.ylabel('Frecuencia')\n",
        "    plt.title('Comparación de Codones Antes y Después de la Edición Genética')\n",
        "    plt.legend()\n",
        "    plt.tight_layout()\n",
        "    plt.show()\n",
        "\n",
        "    # Visualización del contenido de GC\n",
        "    gc_content_before = GC(Seq(old_seq))\n",
        "    gc_content_after = GC(Seq(new_seq))\n",
        "    plt.figure()\n",
        "    plt.bar([\"Antes\", \"Después\"], [gc_content_before, gc_content_after], color=['blue', 'orange'])\n",
        "    plt.ylabel('Contenido de GC (%)')\n",
        "    plt.title('Contenido de GC Antes y Después de la Edición Genética')\n",
        "    plt.show()\n",
        "\n",
        "def process_files(best_guides_df, all_guides_df):\n",
        "    if best_guides_df.empty:\n",
        "        print(\"Error: El DataFrame de mejores guías está vacío.\")\n",
        "        return\n",
        "\n",
        "    for file_name in best_guides_df['File'].unique():\n",
        "        guide_df = best_guides_df[best_guides_df['File'] == file_name]\n",
        "        if guide_df.empty:\n",
        "            print(f\"Error: No se encontraron guías para el archivo '{file_name}'.\")\n",
        "            continue\n",
        "\n",
        "        guide = guide_df['Guide'].iloc[0]\n",
        "\n",
        "        try:\n",
        "            with open(file_name, \"r\") as fasta_file:\n",
        "                sequence = str(SeqIO.read(fasta_file, \"fasta\").seq)\n",
        "                edited_sequence, mutations = transform_bacteria(guide, sequence)\n",
        "                analyze_and_visualize(file_name, sequence, edited_sequence, mutations)\n",
        "        except FileNotFoundError:\n",
        "            print(f\"Error: Archivo '{file_name}' no encontrado en el sistema de archivos.\")\n",
        "        except Exception as e:\n",
        "            print(f\"Error al procesar el archivo '{file_name}': {e}\")\n",
        "\n",
        "# Ejemplo de uso\n",
        "#fasta_files = ['U12441.2.fasta', 'AY509004.fasta']  # Lista de tus archivos FASTA\n",
        "#all_guides_df = process_fasta_files(fasta_files)\n",
        "#best_guides_df = select_best_guides_per_file(all_guides_df)\n",
        "#process_files(best_guides_df, fasta_files)\n"
      ],
      "metadata": {
        "id": "ePFEkFYSlNRa"
      },
      "id": "ePFEkFYSlNRa",
      "execution_count": 32,
      "outputs": []
    }
  ],
  "metadata": {
    "colab": {
      "provenance": [],
      "machine_shape": "hm"
    },
    "language_info": {
      "name": "python"
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "widgets": {
      "application/vnd.jupyter.widget-state+json": {
        "fab05358eeda49878e4c60ebb7c5d3e4": {
          "model_module": "@jupyter-widgets/output",
          "model_name": "OutputModel",
          "model_module_version": "1.0.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/output",
            "_model_module_version": "1.0.0",
            "_model_name": "OutputModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/output",
            "_view_module_version": "1.0.0",
            "_view_name": "OutputView",
            "layout": "IPY_MODEL_47a736ff96d04839a63d4707b8dbc9f0",
            "msg_id": "",
            "outputs": []
          }
        },
        "47a736ff96d04839a63d4707b8dbc9f0": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "84fcecaa4f194245aaef610ec099655c": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "ButtonModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "ButtonModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "ButtonView",
            "button_style": "",
            "description": "Instalar Librerías",
            "disabled": false,
            "icon": "",
            "layout": "IPY_MODEL_e2bba9d4197c4c84ba35e5944acbdaca",
            "style": "IPY_MODEL_de922e72fe0d4bec8cb991b82fd66729",
            "tooltip": ""
          }
        },
        "e2bba9d4197c4c84ba35e5944acbdaca": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "de922e72fe0d4bec8cb991b82fd66729": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "ButtonStyleModel",
          "model_module_version": "1.5.0",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "ButtonStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "button_color": null,
            "font_weight": ""
          }
        },
        "3f0c138212804cd2b041519ac1ceea39": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "VBoxModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "VBoxModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "VBoxView",
            "box_style": "",
            "children": [
              "IPY_MODEL_8c8215497630446e8149ad684a1aed7c",
              "IPY_MODEL_c554ae938f174612b2f552bae358fd1f"
            ],
            "layout": "IPY_MODEL_c7aa1fcc766e4adcb7225b6e9e307ff9"
          }
        },
        "8c8215497630446e8149ad684a1aed7c": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "LabelModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "LabelModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "LabelView",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_19c877fcda3c4c6b80f69f01df225fb8",
            "placeholder": "​",
            "style": "IPY_MODEL_a3660ac834fd41238cdae2566b7ebfc1",
            "value": "Instalación completada."
          }
        },
        "c554ae938f174612b2f552bae358fd1f": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "IntProgressModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "IntProgressModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "ProgressView",
            "bar_style": "info",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_a7ee45cb347b4689b233220643695c73",
            "max": 17,
            "min": 0,
            "orientation": "horizontal",
            "style": "IPY_MODEL_41505775e0284e18810c991d290f047b",
            "value": 17
          }
        },
        "c7aa1fcc766e4adcb7225b6e9e307ff9": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "19c877fcda3c4c6b80f69f01df225fb8": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "a3660ac834fd41238cdae2566b7ebfc1": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "DescriptionStyleModel",
          "model_module_version": "1.5.0",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DescriptionStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "description_width": ""
          }
        },
        "a7ee45cb347b4689b233220643695c73": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "41505775e0284e18810c991d290f047b": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "ProgressStyleModel",
          "model_module_version": "1.5.0",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "ProgressStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "bar_color": null,
            "description_width": ""
          }
        }
      }
    }
  },
  "nbformat": 4,
  "nbformat_minor": 5
}