{"cells":[{"cell_type":"markdown","id":"dbaae3b4","metadata":{"id":"dbaae3b4"},"source":["\n","# Análisis de Secuencias Nucleotídicas de Integronas de Shewanella\n","\n","Este cuaderno realiza la descarga y análisis de secuencias nucleotídicas de integronas de Shewanella utilizando Biopython.\n"]},{"cell_type":"markdown","source":["\n","\n","Este código descarga el dataset de secuencias de integrasas de Shewanella desde el NCBI. Luego, realiza un alineamiento múltiple de las secuencias para identificar las regiones conservadas. A continuación, selecciona un objetivo blanco para las guías CRISPR/Cas9. Finalmente, diseña las guías CRISPR/Cas9 y realiza la transformación de las bacterias de Shewanella que albergan integrones.\n","\n","Para verificar la edición mediante secuenciación, se pueden diseñar primers específicos para la región editada. El código anterior muestra cómo diseñar primers para la región editada del gen de la integrasa.\n","\n","\n","\n","**1. Breve introducción**\n","\n","Los integrones son plataformas genéticas que se encuentran comúnmente en bacterias. Su función principal es la captura y la expresión de genes, los cuales están especialmente asociados con la resistencia a antibióticos. La tecnología CRISPR ha demostrado ser una herramienta poderosa para la edición genética precisa.\n","\n","**2. Hipótesis de Trabajo y Resultados Esperados**\n","\n","La hipótesis del proyecto es que el uso de CRISPR/Cas9 específicamente dirigido al gen de la integrasa en integrones de bacterias del género Shewanella puede resultar en la interrupción o desactivación de la función de recombinación de ADN de la integrasa. Se hipotetiza que al modificar el gen de la integrasa, se reducirá la capacidad de los integrones para capturar y expresar genes asociados con la resistencia a antibióticos en las bacterias de este género.\n","\n","Los resultados esperados del proyecto son:\n","\n","* Demostrar la capacidad de la tecnología CRISPR/Cas9 para dirigirse con precisión al gen de la integrasa de los integrones encontrados en Shewanella, evidenciando la capacidad de editar este gen de manera específica.\n","* Observar una disminución en la capacidad de los integrones modificados para realizar la recombinación del ADN, demostrando así la influencia de la edición genética en la función de la integrasa.\n","* Observar una reducción en la expresión de genes asociados con la resistencia a antibióticos en las bacterias de Shewanella portadoras de integrones editados.\n"],"metadata":{"id":"xGj2f8XsMzKS"},"id":"xGj2f8XsMzKS"},{"cell_type":"code","source":["from Bio import SeqIO\n","from Bio.Seq import Seq\n","\n","# Supongamos que ya tienes los archivos FASTA descargados\n","\n","# Función para modelar la edición de integrones en Shewanella usando CRISPR/Cas9\n","def edit_integrons(accession_numbers):\n","    for accession_number in accession_numbers:\n","        # Leer la secuencia de los integrones\n","        with open(accession_number + \".fasta\", \"r\") as file:\n","            integron_sequence = SeqIO.read(file, \"fasta\")\n","\n","        # Diseñar las guías CRISPR/Cas9\n","        guides = design_guides(integron_sequence.seq)\n","\n","        # Transformar las bacterias de Shewanella con el sistema CRISPR/Cas9\n","        transform_bacteria(guides)\n","\n","        # Verificar la edición mediante secuenciación\n","        verify_editing(accession_number)\n","\n","# Función para diseñar las guías CRISPR/Cas9 (implementación conceptual)\n","\n","def design_guides(sequence):\n","    # Identificar regiones conservadas y posibles sitios de corte\n","    conserved_regions = find_conserved_regions(sequence)\n","    targets = evaluate_target_sites(conserved_regions)\n","    guides = design_guides_from_targets(targets)\n","    return guides\n","\n","# Función para encontrar regiones conservadas (implementación conceptual)\n","def find_conserved_regions(sequence):\n","    # Aquí se podría usar un enfoque bioinformático para identificar regiones conservadas\n","    return [\"region_conservada_1\", \"region_conservada_2\"]\n","\n","# Función para evaluar posibles sitios de corte (implementación conceptual)\n","def evaluate_target_sites(conserved_regions):\n","    # Aquí se evaluarían las regiones conservadas para identificar posibles sitios de corte CRISPR\n","    return [\"sitio_corte_1\", \"sitio_corte_2\"]\n","\n","# Función para diseñar guías a partir de objetivos (implementación conceptual)\n","def design_guides_from_targets(targets):\n","    # Aquí se diseñarían las guías CRISPR basadas en los sitios de corte identificados\n","    return [\"guía_1\", \"guía_2\"]\n","\n","# Función para transformar bacterias (implementación conceptual)\n","def transform_bacteria(guides):\n","    # Aquí se implementaría el proceso de transformación de Shewanella con CRISPR/Cas9\n","    print(\"Transformando bacterias con guías CRISPR/Cas9\")\n","\n","# Función para verificar la edición (implementación conceptual)\n","def verify_editing(accession_number):\n","    # Aquí se diseñarían primers y se realizaría secuenciación para verificar la edición genética\n","    print(f\"Verificando la edición del gen en {accession_number}\")\n","\n","# Ejecución de las funciones\n","edit_integrons(accession_numbers)\n"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"Lgv7s_oHRQeP","executionInfo":{"status":"ok","timestamp":1700767934441,"user_tz":180,"elapsed":462,"user":{"displayName":"rafael perez","userId":"12990849696245978067"}},"outputId":"d303d0ce-5b13-44da-9eb8-c29753c45a49"},"id":"Lgv7s_oHRQeP","execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["Transformando bacterias con guías CRISPR/Cas9\n","Verificando la edición del gen en NC_016848.1\n","Transformando bacterias con guías CRISPR/Cas9\n","Verificando la edición del gen en NC_027179.1\n","Transformando bacterias con guías CRISPR/Cas9\n","Verificando la edición del gen en NC_043643.1\n","Transformando bacterias con guías CRISPR/Cas9\n","Verificando la edición del gen en NC_048237.1\n","Transformando bacterias con guías CRISPR/Cas9\n","Verificando la edición del gen en NC_052996.1\n","Transformando bacterias con guías CRISPR/Cas9\n","Verificando la edición del gen en NC_053976.1\n","Transformando bacterias con guías CRISPR/Cas9\n","Verificando la edición del gen en NC_058644.1\n","Transformando bacterias con guías CRISPR/Cas9\n","Verificando la edición del gen en NC_059384.1\n","Transformando bacterias con guías CRISPR/Cas9\n","Verificando la edición del gen en NC_062436.1\n","Transformando bacterias con guías CRISPR/Cas9\n","Verificando la edición del gen en NC_064046.1\n","Transformando bacterias con guías CRISPR/Cas9\n","Verificando la edición del gen en NC_069283.1\n","Transformando bacterias con guías CRISPR/Cas9\n","Verificando la edición del gen en NC_073197.1\n","Transformando bacterias con guías CRISPR/Cas9\n","Verificando la edición del gen en NC_074563.1\n","Transformando bacterias con guías CRISPR/Cas9\n","Verificando la edición del gen en NC_076980.1\n","Transformando bacterias con guías CRISPR/Cas9\n","Verificando la edición del gen en NC_079963.1\n"]}]},{"cell_type":"code","source":["#\n","def transform_bacteria(guides, sequence):\n","    mutations = []\n","    for guide in guides:\n","        if random.random() < 0.5:  # 50% de posibilidades de una edición exitosa\n","            start = sequence.find(guide)\n","            if start != -1:\n","                end = start + len(guide)\n","                mutation_type = random.choice([\"deletion\", \"insertion\", \"substitution\"])\n","                if mutation_type == \"deletion\":\n","                    sequence = sequence[:start] + sequence[end:]\n","                    mutations.append((mutation_type, start, sequence[start:end], \"\"))\n","                elif mutation_type == \"insertion\":\n","                    insertion_seq = \"ATCG\"\n","                    sequence = sequence[:start] + insertion_seq + sequence[start:]\n","                    mutations.append((mutation_type, start, \"\", insertion_seq))\n","                elif mutation_type == \"substitution\":\n","                    substitution_seq = \"ATCG\"\n","                    sequence = sequence[:start] + substitution_seq + sequence[end:]\n","                    mutations.append((mutation_type, start, sequence[start:end], substitution_seq))\n","    return sequence, mutations\n","\n","def verify_editing(accession_number, original_sequence, edited_sequence, mutations):\n","    if original_sequence != edited_sequence:\n","        print(f\"Edición genética verificada para {accession_number}\")\n","        print(\"Detalle de las mutaciones:\")\n","        for mutation in mutations:\n","            mutation_type, position, original, new = mutation\n","            print(f\"{mutation_type.capitalize()} en posición {position}: Original '{original}', Nuevo '{new}'\")\n","\n","        # Representación gráfica de las ediciones\n","        print(\"\\nRepresentación Gráfica de las Ediciones:\")\n","        graphical_representation(original_sequence, edited_sequence, mutations)\n","        return edited_sequence\n","    else:\n","        print(f\"No se detectó edición genética para {accession_number}\")\n","        return None\n","\n","def graphical_representation(original_seq, edited_seq, mutations):\n","    print(\"Original: \" + original_seq[:50] + \"...\" + original_seq[-50:])\n","    print(\"Editada : \" + edited_seq[:50] + \"...\" + edited_seq[-50:])\n","    for mutation in mutations:\n","        mutation_type, position, original, new = mutation\n","        print(f\"{'^' * len(original)} {mutation_type.capitalize()} en posición {position}\")\n","\n","# Ejemplo de uso\n","accession_number = \"NC_016848.1\"\n","with open(accession_number + \".fasta\", \"r\") as file:\n","    original_sequence = SeqIO.read(file, \"fasta\").seq\n","\n","guides = [\"GTTACG\", \"CGTACG\"]  # Ejemplo de guías CRISPR\n","edited_sequence, mutations = transform_bacteria(guides, original_sequence)\n","\n","edited_seq_record = verify_editing(accession_number, original_sequence, edited_sequence, mutations)\n","if edited_seq_record:\n","    SeqIO.write(SeqRecord(Seq(edited_seq_record), id=accession_number), f\"{accession_number}_edited.fasta\", \"fasta\")"],"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":367},"id":"b_ojOgi5ZAEc","executionInfo":{"status":"error","timestamp":1700767934442,"user_tz":180,"elapsed":18,"user":{"displayName":"rafael perez","userId":"12990849696245978067"}},"outputId":"eebf1ff4-3c99-40c3-f077-854f47062e8f"},"id":"b_ojOgi5ZAEc","execution_count":null,"outputs":[{"output_type":"error","ename":"NameError","evalue":"ignored","traceback":["\u001b[0;31m---------------------------------------------------------------------------\u001b[0m","\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)","\u001b[0;32m<ipython-input-9-cde829d789a5>\u001b[0m in \u001b[0;36m<cell line: 52>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     50\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     51\u001b[0m \u001b[0mguides\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m\"GTTACG\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"CGTACG\"\u001b[0m\u001b[0;34m]\u001b[0m  \u001b[0;31m# Ejemplo de guías CRISPR\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 52\u001b[0;31m \u001b[0medited_sequence\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmutations\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtransform_bacteria\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mguides\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0moriginal_sequence\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     53\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     54\u001b[0m \u001b[0medited_seq_record\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mverify_editing\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0maccession_number\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0moriginal_sequence\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0medited_sequence\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmutations\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m<ipython-input-9-cde829d789a5>\u001b[0m in \u001b[0;36mtransform_bacteria\u001b[0;34m(guides, sequence)\u001b[0m\n\u001b[1;32m      3\u001b[0m     \u001b[0mmutations\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m     \u001b[0;32mfor\u001b[0m \u001b[0mguide\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mguides\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 5\u001b[0;31m         \u001b[0;32mif\u001b[0m \u001b[0mrandom\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrandom\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m<\u001b[0m \u001b[0;36m0.5\u001b[0m\u001b[0;34m:\u001b[0m  \u001b[0;31m# 50% de posibilidades de una edición exitosa\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      6\u001b[0m             \u001b[0mstart\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msequence\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfind\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mguide\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      7\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mstart\u001b[0m \u001b[0;34m!=\u001b[0m \u001b[0;34m-\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;31mNameError\u001b[0m: name 'random' is not defined"]}]},{"cell_type":"code","source":["import random\n","from Bio.Seq import Seq\n","from Bio.SeqRecord import SeqRecord\n","from Bio import SeqIO\n","import matplotlib.pyplot as plt\n","\n","# Función para transformar bacterias simulando ediciones CRISPR/Cas9\n","def transform_bacteria(guides, sequence):\n","    mutations = []\n","    for guide in guides:\n","        if random.random() < 0.5:  # 50% de posibilidades de una edición exitosa\n","            start = sequence.find(guide)\n","            if start != -1:\n","                end = start + len(guide)\n","                mutation_type = random.choice([\"deletion\", \"insertion\", \"substitution\"])\n","                if mutation_type == \"deletion\":\n","                    sequence = sequence[:start] + sequence[end:]\n","                    mutations.append((mutation_type, start, sequence[start:end], \"\"))\n","                elif mutation_type == \"insertion\":\n","                    insertion_seq = \"ATCG\"\n","                    sequence = sequence[:start] + insertion_seq + sequence[start:]\n","                    mutations.append((mutation_type, start, \"\", insertion_seq))\n","                elif mutation_type == \"substitution\":\n","                    substitution_seq = \"ATCG\"\n","                    sequence = sequence[:start] + substitution_seq + sequence[end:]\n","                    mutations.append((mutation_type, start, sequence[start:end], substitution_seq))\n","    return sequence, mutations\n","\n","# Función para verificar la edición genética\n","def verify_editing(accession_number, original_sequence, edited_sequence, mutations):\n","    if original_sequence != edited_sequence:\n","        print(f\"Edición genética verificada para {accession_number}\")\n","        summarize_mutations(mutations)\n","        graphical_representation(original_sequence, edited_sequence, mutations)\n","        plot_sequence_differences(original_sequence, edited_sequence, mutations)\n","        return edited_sequence\n","    else:\n","        print(f\"No se detectó edición genética para {accession_number}\")\n","        return None\n","\n","# Función para resumir las mutaciones\n","def summarize_mutations(mutations):\n","    print(f\"\\nTotal de mutaciones: {len(mutations)}\")\n","    count_deletion = sum(1 for m in mutations if m[0] == \"deletion\")\n","    count_insertion = sum(1 for m in mutations if m[0] == \"insertion\")\n","    count_substitution = sum(1 for m in mutations if m[0] == \"substitution\")\n","\n","    print(f\"Deleciones: {count_deletion}\")\n","    print(f\"Inserciones: {count_insertion}\")\n","    print(f\"Sustituciones: {count_substitution}\")\n","\n","# Función para representación gráfica de las ediciones\n","def graphical_representation(original_seq, edited_seq, mutations):\n","    print(\"\\nOriginal: \" + original_seq[:50] + \"...\" + original_seq[-50:])\n","    print(\"Editada : \" + edited_seq[:50] + \"...\" + edited_seq[-50:])\n","    for mutation in mutations:\n","        mutation_type, position, original, new = mutation\n","        print(f\"{'^' * len(original)} {mutation_type.capitalize()} en posición {position}\")\n","\n","# Función para graficar las diferencias de secuencias\n","def plot_sequence_differences(original_seq, edited_seq, mutations):\n","    positions = [pos for _, pos, _, _ in mutations]\n","    mutation_types = [mut_type for mut_type, _, _, _ in mutations]\n","\n","    plt.figure(figsize=(10, 3))\n","    plt.scatter(positions, [1] * len(positions), c='red', label='Mutaciones')\n","    for pos, mut_type in zip(positions, mutation_types):\n","        plt.annotate(mut_type, (pos, 1), textcoords=\"offset points\", xytext=(0,10), ha='center')\n","\n","    plt.title('Distribución de Mutaciones en la Secuencia')\n","    plt.xlabel('Posición en la Secuencia')\n","    plt.yticks([])\n","    plt.legend()\n","    plt.show()\n","\n","# Ejemplo de uso\n","accession_number = \"NC_016848.1\"\n","with open(accession_number + \".fasta\", \"r\") as file:\n","    original_sequence = SeqIO.read(file, \"fasta\").seq\n","\n","guides = [\"GTTACG\", \"CGTACG\"]  # Ejemplo de guías CRISPR\n","edited_sequence, mutations = transform_bacteria(guides, original_sequence)\n","\n","edited_seq_record = verify_editing(accession_number, original_sequence, edited_sequence, mutations)\n","if edited_seq_record:\n","    SeqIO.write(SeqRecord(Seq(edited_seq_record), id=accession_number), f\"{accession_number}_edited.fasta\", \"fasta\")\n"],"metadata":{"id":"AZYXt7j0dwC-"},"id":"AZYXt7j0dwC-","execution_count":null,"outputs":[]}],"metadata":{"colab":{"provenance":[],"machine_shape":"hm"},"language_info":{"name":"python"},"kernelspec":{"name":"python3","display_name":"Python 3"}},"nbformat":4,"nbformat_minor":5}