{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "YAxgZo4Y4Zx3"
      },
      "source": [
        "#Trabajo Final Grupo Nº 2\n",
        "\n",
        "Rafael Perez\trafaelperezctes@gmail.com\n",
        "Olinca Ayala\tolincayalan88@gmail.com\n",
        "###@Created on Dec Sat 2 16:57:20 2023\n",
        "###@author: Grupo Nº 2"
      ],
      "id": "YAxgZo4Y4Zx3"
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "dbaae3b4"
      },
      "source": [
        "\n",
        "# Modulo Vision\n",
        "\n",
        "Este cuaderno realiza la descarga y análisis de secuencias nucleotídicas de integronas de Shewanella utilizando Biopython.\n"
      ],
      "id": "dbaae3b4"
    },
    {
      "cell_type": "code",
      "source": [
        "# @title Instalación de librerias\n",
        "import subprocess\n",
        "import ipywidgets as widgets\n",
        "from IPython.display import display\n",
        "import pkg_resources\n",
        "\n",
        "def instalar_librerias():\n",
        "    librerias = [\n",
        "        \"import_ipynb\",\n",
        "        \"nbconvert\",\n",
        "        \"ipywidgets\",\n",
        "        \"biopython\",\n",
        "        \"pandas\",\n",
        "        \"numpy\",\n",
        "        \"matplotlib\",\n",
        "        \"tqdm\",\n",
        "        \"odfpy\",\n",
        "        \"requests\",\n",
        "        \"gdown\",\n",
        "        \"mafft\",\n",
        "        \"clustalo\",\n",
        "        \"muscle\",\n",
        "        \"fasttree\",\n",
        "        \"weblogo\",\n",
        "        \"logomaker\",\n",
        "\n",
        "    ]\n",
        "\n",
        "    barra_progreso = widgets.IntProgress(\n",
        "        value=0,\n",
        "        min=0,\n",
        "        max=len(librerias),\n",
        "        bar_style='info',\n",
        "        orientation='horizontal'\n",
        "    )\n",
        "\n",
        "    etiqueta = widgets.Label(value='Iniciando instalación...')\n",
        "    display(widgets.VBox([etiqueta, barra_progreso]))\n",
        "\n",
        "    # Instalación de fasttree y mafft con apt-get (solo ejecutar en Colab)\n",
        "    subprocess.run([\"apt-get\", \"install\", \"fasttree\", \"mafft\", \"ghostscript\"])\n",
        "\n",
        "    for libreria in librerias:\n",
        "        try:\n",
        "            pkg_resources.require(libreria)\n",
        "            etiqueta.value = f\"{libreria} ya está instalado.\"\n",
        "        except pkg_resources.DistributionNotFound:\n",
        "            etiqueta.value = f\"Instalando: {libreria}...\"\n",
        "            subprocess.run([\"pip\", \"install\", libreria])\n",
        "        barra_progreso.value += 1\n",
        "\n",
        "    etiqueta.value = 'Instalación completada.'\n",
        "    # importo todas las librerias\n",
        "    importarLibrerias()\n",
        "# Crear botón\n",
        "boton_instalar = widgets.Button(description='Instalar Librerías')\n",
        "\n",
        "# Crear área de salida\n",
        "output = widgets.Output()\n",
        "display(output)\n",
        "\n",
        "# Enlazar el botón con la función de instalación\n",
        "boton_instalar.on_click(lambda b: instalar_librerias())\n",
        "\n",
        "# Mostrar botón\n",
        "#display(boton_instalar)\n",
        "\n",
        "#instalar_librerias()\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 49,
          "referenced_widgets": [
            "e7aca3eb0bff46cebf8be71c55539c7f",
            "9e7c276c16d74add9e509c14e774f024",
            "3ab74832422e4ba2a0daf0d31b96d320",
            "7a2de3dbfc60491489b420875c7ed6b7",
            "99eeba0fa61841ed9f8f79af0b50061d"
          ]
        },
        "id": "rlMGb_QOAR9K",
        "outputId": "ce078506-43e2-4e62-9ecd-5231f57da897"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "Output()"
            ],
            "application/vnd.jupyter.widget-view+json": {
              "version_major": 2,
              "version_minor": 0,
              "model_id": "e7aca3eb0bff46cebf8be71c55539c7f"
            }
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "Button(description='Instalar Librerías', style=ButtonStyle())"
            ],
            "application/vnd.jupyter.widget-view+json": {
              "version_major": 2,
              "version_minor": 0,
              "model_id": "3ab74832422e4ba2a0daf0d31b96d320"
            }
          },
          "metadata": {}
        }
      ],
      "id": "rlMGb_QOAR9K"
    },
    {
      "cell_type": "code",
      "source": [
        "def main():\n",
        "  instalar_librerias()\n",
        "#main()"
      ],
      "metadata": {
        "id": "rY2HquGAyIa0"
      },
      "execution_count": null,
      "outputs": [],
      "id": "rY2HquGAyIa0"
    },
    {
      "cell_type": "code",
      "source": [
        "# -*- coding: utf-8 -*-\n",
        "\"\"\"\n",
        "Created on Tue Oct 19 14:17:44 2021\n",
        "\n",
        "@author: android-2d\n",
        "\"\"\"\n",
        "\n",
        "from Bio.PDB.MMCIFParser import MMCIFParser\n",
        "from Bio.PDB.PDBParser import PDBParser\n",
        "from Bio.PDB.MMCIF2Dict import MMCIF2Dict\n",
        "import warnings\n",
        "warnings.filterwarnings(\"ignore\")\n",
        "\n",
        "def LeerPDB(structure_id,filename):\n",
        "    parser = PDBParser(PERMISSIVE=1)\n",
        "    structure = parser.get_structure(structure_id, filename)\n",
        "    return structure\n",
        "\n",
        "def LeerCIF(filename):\n",
        "    mmcif_dict = MMCIF2Dict(filename)\n",
        "    for value,key in mmcif_dict.items():\n",
        "        print(value,key)\n",
        "    return mmcif_dict\n",
        "\n",
        "def Graficar(structure):\n",
        "    import nglview as nv\n",
        "    view = nv.show_biopython(structure)\n",
        "    view.render_image()\n",
        "    view._display_image()\n",
        "    view\n",
        "    return view\n",
        "    #view2.clear_representations()\n",
        "    #view2.add_representation('cartoon', selection='protein', color='blue')\n",
        "    #view2.add_representation('licorice', selection='not hydrogen')\n",
        "\n",
        "def SelecionePDB(lista_pdb):\n",
        "    ingreso = int(input(\"Ingrese el numero del PDB a graficar: \"))\n",
        "    if ingreso!=0:\n",
        "        salida = lista_pdb[ingreso-1]\n",
        "    else:\n",
        "        salida=0\n",
        "    return salida\n",
        "\n",
        "def ListarPDB():\n",
        "    import os\n",
        "    import time\n",
        "    actual_dir = os.getcwd()\n",
        "    PDB_dir= actual_dir+\"\\\\BD\\\\PDB\\\\\"\n",
        "    contenido = os.listdir(PDB_dir)\n",
        "    c=1\n",
        "    lista_pdb=[]\n",
        "    print(\"Se han encotrado los siguientes archivos PDB:\")\n",
        "    for archivo in contenido:\n",
        "        if archivo.endswith('.pdb'):\n",
        "            print(c,\":\",archivo)\n",
        "            c+=1\n",
        "            lista_pdb.append(PDB_dir+archivo)\n",
        "\n",
        "    time.sleep(1)\n",
        "    pdb_elegido=SelecionePDB(lista_pdb)\n",
        "    return pdb_elegido"
      ],
      "metadata": {
        "id": "I7EAdMp9l38E"
      },
      "id": "I7EAdMp9l38E",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "def main():\n",
        "        structure_id = 'miPDB'\n",
        "        filename = ListarPDB()\n",
        "        structure = LeerPDB(structure_id,filename)\n",
        "        view=Graficar(structure)\n",
        "        return view\n",
        "\n",
        "main()"
      ],
      "metadata": {
        "id": "WqxFHKrxl8b7"
      },
      "id": "WqxFHKrxl8b7",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#!/usr/bin/env python\n",
        "# coding: utf-8\n",
        "\n",
        "# # TP6 Grupo Nº 5\n",
        "# Perez Ernesto Rafael rafaelperezctes@gmail.com Sofia Erdozain sofierdozain@gmail.com\n",
        "# Created on Wed Nov  3 14:02:59 2021\n",
        "\n",
        "# @author: Grupo Nº 5\n",
        "\n",
        "\n",
        "import sys\n",
        "import Alineamientos\n",
        "import CargaBD\n",
        "import Main as principal\n",
        "import Utils\n",
        "\n",
        "proteinas = CargaBD.proteinas\n",
        "from Bio.Blast import NCBIWWW, NCBIXML\n",
        "from Bio import SeqIO\n",
        "from io import StringIO\n",
        "\n",
        "def buscarBD(archivo):\n",
        "    f_record = next(SeqIO.parse(archivo, \"fasta\"))\n",
        "    # busco en la BD\n",
        "    print(\"corriendo blast recuperando datos...\")\n",
        "    result_handle = NCBIWWW.qblast(\"blastp\", \"pdb\", f_record.format(\"fasta\"))\n",
        "    # guardo resultados\n",
        "    qblast_out = 'm_cold_blast.out'\n",
        "    archivo_p = \"BD/\" + qblast_out\n",
        "    with open(archivo_p, \"w\") as save_file:\n",
        "        blast_results = result_handle.read()\n",
        "        save_file.write(blast_results)\n",
        "    result_handle.close()\n",
        "    return blast_results\n",
        "\n",
        "def AnalizarResultados(blast_results, nombre):\n",
        "    print(\"Analizando los resultados y extrayendo información...\")\n",
        "    # opción 1 - abre el archivo guardado para analizarlo\n",
        "    # opción 2 - crear un identificador a partir de la cadena y analizarlo\n",
        "    string_result_handle = StringIO(blast_results)\n",
        "    b_record = NCBIXML.read(string_result_handle)\n",
        "    # ahora obtengo la información de alineación\n",
        "    # para todos los valores e mayores que algún umbral\n",
        "    E_VALUE_THRESH = 0.5\n",
        "    lista_blast = []\n",
        "    print(\"Se encontraron \", len(b_record.alignments), \"estructuras asociadas a \", nombre)\n",
        "    c = 1\n",
        "    for alignment in b_record.alignments:\n",
        "        for hsp in alignment.hsps:\n",
        "            # if hsp.expect < E_VALUE_THRESH:\n",
        "            print()\n",
        "            print(\"Proteina \", c, \" ID:\", alignment.accession)\n",
        "            lista_blast.append(str(alignment.accession))\n",
        "            print(\"****Alineacion****\")\n",
        "            print(\"secuencia: %s\" % alignment.title)\n",
        "            print(\"longitud: %i\" % alignment.length)\n",
        "            print(\"e valor: %f\" % hsp.expect)\n",
        "            print(hsp.query[0:75] + \"...\")\n",
        "            print(hsp.match[0:75] + \"...\")\n",
        "            print(hsp.sbjct[0:75] + \"...\")\n",
        "        c += 1\n",
        "    return lista_blast\n",
        "\n",
        "\n",
        "def SubMenu2(lista_blast):\n",
        "    print()\n",
        "    print(\"### Modulo: Estructuras NCBI ###\")\n",
        "    print()\n",
        "    indice = int(input(\"Elija el numero de la estructura que desea descargar - 0 para salir: \"))\n",
        "    if indice == 0:\n",
        "        SubMenu1()\n",
        "    else:\n",
        "        from Bio.PDB import PDBList\n",
        "        pdbl = PDBList()\n",
        "        lista_pdb = [lista_blast[indice - 1][0:4]]  # debo buscar solo las 4 letras iniciales\n",
        "        pdbl.download_pdb_files(lista_pdb, obsolete=False, pdir=\"BD/PDB/\", file_format=\"pdb\", overwrite=True)\n",
        "        # pdbl.retrieve_pdb_file('1ms5', file_format=\"pdb\")\n",
        "        # https://files.rcsb.org/download/1MS5.pdb\n",
        "\n",
        "\n",
        "def SubMenu1():\n",
        "    print(\"### Modulo: Estructuras NCBI ###\")\n",
        "    print()\n",
        "    Utils.ListarProteinas()\n",
        "    proteinas_lista = list(CargaBD.proteinas)\n",
        "    indice = int(input(\"Elija el numero de la proteina para analizar: \"))\n",
        "    nombre = str(proteinas_lista[indice - 1])\n",
        "    archivo = \"BD/\" + nombre + \".fasta\"\n",
        "    return archivo, nombre\n",
        "\n",
        "\n",
        "def Visualizar():\n",
        "    import subprocess\n",
        "    from subprocess import PIPE\n",
        "    print(\n",
        "        \"Se ejecutara Jupyter-netbook en su navegador y se abrira una consola (Cierre la consola de Jupyter para volver al menu )\")\n",
        "    comando = \"jupyter notebook EstructurasNCBI-visualizador.ipynb\"\n",
        "    result = subprocess.Popen(comando)\n",
        "    result.communicate()\n",
        "    return 1\n",
        "\n",
        "\n",
        "def Menu():\n",
        "    print()\n",
        "    print(\"### Modulo: Estructuras NCBI###\")\n",
        "    print()\n",
        "    print(\"Que desea realizar:\")\n",
        "    print(\"1- Buscar estructura para Proteinas almacenadas\")\n",
        "    print(\"2- Visualizar proiteinas\")\n",
        "    print(\"3- Alinear (Beta)\")\n",
        "    print(\"4- Volver al menu principal\")\n",
        "    print(\"5- Salir\")\n",
        "    CargaBD.LeerInicio()\n",
        "    opcion_principal = int(input(\"Ingrese una opción: \"))\n",
        "    if opcion_principal == 1:\n",
        "        archivo, nombre = SubMenu1()\n",
        "        blast_results = buscarBD(archivo)\n",
        "        lista_blast = AnalizarResultados(blast_results, nombre)\n",
        "        SubMenu2(lista_blast)\n",
        "        Menu()\n",
        "    if opcion_principal == 2:\n",
        "        Visualizar()\n",
        "        Menu()\n",
        "    if opcion_principal == 3:\n",
        "        Utils.ListarGenomas()\n",
        "        Menu()\n",
        "    if opcion_principal == 4:\n",
        "        principal.Menu()\n",
        "    if opcion_principal == 5:\n",
        "        print(\"Gracias por utilizar BIOG5\")\n",
        "        sys.exit()\n",
        "# Menu()"
      ],
      "metadata": {
        "id": "lnwJT3Avl9VF"
      },
      "id": "lnwJT3Avl9VF",
      "execution_count": null,
      "outputs": []
    }
  ],
  "metadata": {
    "colab": {
      "provenance": [],
      "machine_shape": "hm"
    },
    "language_info": {
      "name": "python"
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "widgets": {
      "application/vnd.jupyter.widget-state+json": {}
    }
  },
  "nbformat": 4,
  "nbformat_minor": 5
}